libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -250,22 +265,312 @@ internal class NativeLoginManager(
+        otp: String,
+        otpIdentifier: String,
+        otpVerificationMethod: OtpVerificationMethod
+    ): NativeLoginResult {
+        // Validate parameters.
+        val trimmedOtp = otp.trim()
+
+        // Generate code verifier and code challenge.
+        val codeVerifier = getRandom128ByteKey()
+        val codeChallenge = getSHA256Hash(codeVerifier)
+
+        // Determine the OTP verification method.
+        otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue
+
+        // Generate the authorization.
+        val authorization = generateColonConcatenatedBase64String(
+            value1 = otpIdentifier,
+            value2 = trimmedOtp
+        )
+
+        // Generate the authorization request headers.
+        val authorizationRequestHeaders = hashMapOf(
+            AUTH_REQUEST_TYPE_HEADER_NAME to AUTH_REQUEST_TYPE_VALUE_PASSWORDLESS_LOGIN,
+            AUTH_VERIFICATION_TYPE_HEADER_NAME to otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue,
+            CONTENT_TYPE_HEADER_NAME to CONTENT_TYPE_VALUE_HTTP_POST,
+            AUTHORIZATION to "$AUTH_AUTHORIZATION_VALUE_BASIC $authorization"
+        )
+
+        // Generate the authorization request body.
+        val authorizationRequestBodyString = createRequestBody(
+            RESPONSE_TYPE to CODE_CREDENTIALS,
+            CLIENT_ID to clientId,
+            REDIRECT_URI to redirectUri,
+            CODE_CHALLENGE to codeChallenge,
+        )
+
+        // Create the authorization request.
+        val authorizationRequest = RestRequest(
+            POST,
+            LOGIN,
+            "$loginUrl$OAUTH_AUTH_PATH",
+            authorizationRequestBodyString,
+            authorizationRequestHeaders,
+        )
+
+        /*
+         * Submit the authorization request, fetch the authorization response,
+         * submit the access token request and return an appropriate login
+         * result.
+         */
+        return submitAccessTokenRequest(
+            authorizationResponse = suspendedRestCall(
+                authorizationRequest
+            ) ?: return UnknownError,
+            codeVerifier = codeVerifier
+        )
+    }
+
+    /**
+     * A data class for the OTP request body.
+     * @param recaptcha The reCAPTCHA token provided by the reCAPTCHA Android
+     * SDK.  This is not used with reCAPTCHA Enterprise
+     * @param recaptchaevent The reCAPTCHA parameters for use with reCAPTCHA
+     * Enterprise
+     * @param username The Salesforce username
+     * @param verificationMethod The OTP verification code's delivery method in
+     * "email" or "sms"
+     */
+    private data class OtpRequestBody(
+        val recaptcha: String?,
+        val recaptchaevent: OtpRequestBodyReCaptchaEvent?,
+        val username: String,
+        val verificationMethod: String
+    ) {
+        // TODO: Inquire regarding modern JSON serialization in MSDK. ECJ20240317
+        fun toJson() = JSONObject().apply {
```

Comment:
@brandonpage or @wmathurin - I took a casual glance around the rest of MSDK and didn't see any newer way of working with JSON beyond json.org.  Is that true? It would be trivial to update this to Kotlin Serialization.